DefaultArgParser: Add generic filter-by option

shreyans800755 opened the pull request:
Closes https://github.com/coala/coala/issues/2509

<!--
Thanks for your contribution!

Please take a quick look at those things down there. They're quite important.
Really! We wrote them for you. Yes you! With utmost care. Read them.
-->

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

shreyans800755 commented:
This PR needs work regarding adding filters related to can_detect, can_fix, etc.
I've created PR just to show the approach before moving ahead.

Here, we can add filtering logic inside FilterImplementation.py to add more options like can_detect, can_fix, etc. Currently I've only added one option for language since its logic was already available in coala.py.
@Makman2 
Please review the approach.

FYI: One more approach I could think was to create abstract class with method to get bears and implement all the filters as different class inheriting from the abstract class. But it would require creation of a new file every time we need to add an filter. So, I did not choose this approach.

Makman2 commented on coalib/coala.py:
```diff
@@ -41,18 +41,35 @@ def main():
-                       console_printer)
-
+            # filter_by_language option should be deprecated soon.
+            # Since it is already covered by filter_by option.
```
log a warning ;)

Makman2 commented on coalib/parsing/DefaultArgParser.py:
```diff
@@ -187,6 +187,10 @@ def default_arg_parser(formatter_class=None):
         help='filters `--show-bears` by the given languages')
 
     outputs_group.add_argument(
+        '--filter-by', nargs='+',
```
normally I would say that this would be good and intuitive, but I'm not sure how it plays well with other arguments. coala takes quite many currently, and I have the feeling that this doesn't work any more:
```
# The args here don't make sense together, but just for what I mean:
coala --filter-by filter1 filter2 --non-interactive --save
```
But I recall that argparse might still handle that properly... could you test that? :)

shreyans800755 replied:
It works fine. Although  need to add `-B` before `--filter-by`.
argparse takes care of that.

Makman2 commented on coalib/parsing/FilterImplementation.py:
```diff
@@ -0,0 +1,7 @@
+def LanguageFilter(languages, log_printer):
```
so we could still do one filter one file, can't we? Actually the more you modularize, the better, maintaining multiple files is not a problem, it's even better than maintaining a single file ;)

Makman2 commented on coalib/parsing/Filter.py:
```diff
@@ -0,0 +1,19 @@
+        return filter in Filter.all_filters
+
+    @staticmethod
+    def get_filtered_bears(filter, args, log_printer):
```
don't use `log_printer`, use `logging` instead. `log_printer` is inofficially deprecated and we are going to remove it step by step in future. The current CLI is just quite complicated and we are all to lazy to do that properly :/

shreyans800755 replied:
Is there any issue taking care of it ? If not, I can file it and also start working on it. IMO, it is going to be time consuming to remove dependency from `log_printer`.

Makman2 replied:
could you search for one? If not then feel free to file one :+1:

shreyans800755 replied:
Already there, https://github.com/coala/coala/issues/3472.

Makman2 commented on coalib/parsing/Filter.py:
```diff
@@ -0,0 +1,19 @@
+    @staticmethod
+    def get_filtered_bears(filter, args, log_printer):
+        if Filter.is_valid_filter(filter):
+            return Filter.all_filters[filter](args, log_printer)
```
what about passing each filter the list of bears? :) Then the filters don't have to look for them themselves, but we have one function collecting them all, and then each filter can process them :)

Makman2 replied:
(together with the filter options)

shreyans800755 replied:
Sounds good.

Makman2 commented:
Idea and direction looks good, you distribute different types of filter across functions/classes/other entities.

> FYI: One more approach I could think was to create abstract class with method to get bears and implement all the filters as different class inheriting from the abstract class. But it would require creation of a new file every time we need to add an filter. So, I did not choose this approach.

The more files, the better to maintain :) As this speaks for a high modularization of your functionality, which is always good. But it doesn't have to be classes, you can also use functions (like you already do with `LanguageFilter`, weird naming convention btw for functions ;D).

shreyans800755 commented:
@Makman2 
This is wip, but I want to confirm the approach. I've done LanguageFilter part. I'll add other filters soon.
Testing is still remaining(which is causing the CI builds failure).

I'm not removing log_printer as of now, as there's separate issue for that, which will take care of it, I believe.(coz I'm working on that issue :)).

Makman2 commented:
Regarding `log_printers`: Just don't use them, saves code and is easier ;)

Makman2 commented on coalib/settings/ConfigurationGathering.py:
```diff
@@ -329,11 +326,6 @@ def get_filtered_bears(languages, log_printer, arg_parser=None):
-        local_bears = filter_section_bears_by_languages(
-            local_bears, languages)
-        global_bears = filter_section_bears_by_languages(
-            global_bears, languages)
```
perfect :+1:

Makman2 commented on coalib/parsing/Filter.py:
```diff
@@ -0,0 +1,22 @@
+
+
+class Filter:
+    all_filters = {'language': language_filter}
```
I think to modularize things more, you should provide a function `registerFilter` where someone can register his own filter object/class/function/etc. Then we can dynamically built a filter with different sets of functionality (though I cannot think of a useful case right now).

This would work for example like https://docs.python.org/3/library/unittest.html#unittest.TestCase.addTypeEqualityFunc

shreyans800755 replied:
I think I'll first complete rest, and then work on this one. This sounds like an interesting feature. But may require a bit more design.

Makman2 replied:
shouldn't be that hard though ;)

Makman2 commented on coalib/coala.py:
```diff
@@ -42,18 +42,39 @@ def main():
+            if args.filter_by:
+                from coalib.parsing.Filter import Filter
+                filter = args.filter_by[0]
+                filter_args = args.filter_by[1:]
```
what about multiple filters? Like
```
--filter-by language cpp --filter-by aspect License
```
or
```
--filter-by language cpp&haskell
```

Makman2 replied:
The syntax can be different, but just as an example for more complex filters^^

Makman2 replied:
Ah and what about aliases? So I can do
```
--filter-by lang cpp
# equivalent to
# --filter-by language cpp
```

shreyans800755 replied:
I had thought of this, but the issue comes with arg parser. 

> --filter-by language cpp --filter-by aspect License

In case of multiple filter-by available, it only yields results for the later filter. I'll try to fix that.

> --filter-by language cpp&haskell

Currently `And` is not available as operator, but if space separated list is provided, it gives bears with `or` operator.
So, for the following:

> coala -B --filter-by language c python

It gives all the bears that can be used of either C or python or both.

As far as `and` operator is concerned, fixing of multiple `--filter-by` will help.

I've not thought of aliases, at the moment. But we just need to add an entry in the map for that. That's fairly easy to do.

Makman2 replied:
> >  --filter-by language cpp --filter-by aspect License
>
> In case of multiple filter-by available, it only yields results for the later filter. I'll try to fix that.

That would be good to have definitely. But actually we can merge the more primitive version first and improve from that. Step by step is always good ;)

okay an "and" operator is anyway not really useful, but it's good that "or" is supported :+1: just an impression that we may want to support multiple operators.

> I've not thought of aliases, at the moment. But we just need to add an entry in the map for that. That's fairly easy to do.

:+1:

shreyans800755 replied:
`or` operator is allowed only within a single filter.
It means we can write `filter-by C or python`but not`filter-by C or filter-by can_fix syntax`. Although I'm not sure later is that useful. It is mainly useful with`and`operators, which I'm a working on to implement.

Makman2 commented on coalib/parsing/filters/__init__.py:
```diff
@@ -0,0 +1,7 @@
+get the list of bears with specific properties.
+
+Example:
+    LanguageFilter: To get the list of bears for provided language/
```
I wouldn't provide an example, I think people know how to handle the package roughly once they see its content ;)

shreyans800755 replied:
Cool

Makman2 commented on coalib/parsing/filters/LanguageFilter.py:
```diff
@@ -0,0 +1,27 @@
+def language_filter(bears, languages, log_printer):
+    local_bears, global_bears = bears
```
do we want to separate local and global bears? I wouldn't do that, especially as in the next-gen core we won't make a distinction any more in narrower sense, as their difference is only the degree in parallelization. The current core has special code for both of them, but the next-gen core won't. The bears do specify themselves the degree of parallelization^^

Makman2 replied:
ah but we need that for `show_bears`... ugh that sucks...

Makman2 replied:
why the heck do we need to display bears differently depending on whether they are a `GlobalBear` or `LocalBear`... maybe we wanna change that @shreyans800755 ? But that's quite a task, so you don't have to ;)

shreyans800755 replied:
Should I add warning log in show bears regarding parameters deprecation ?

Makman2 commented on coalib/parsing/filters/LanguageFilter.py:
```diff
@@ -0,0 +1,27 @@
+        local_bears = filter_section_bears_by_languages(local_bears, languages)
+        global_bears = filter_section_bears_by_languages(
+            global_bears, languages)
+    return local_bears, global_bears
```
I imagine currently that every filter will contain this code, maybe you invoke in the "filter-core" the filter function twice? Once for local and once for global bears? Then people only have to write the filter^^

shreyans800755 commented on coalib/parsing/filters/UseStdinFilter.py:
```diff
@@ -0,0 +1,31 @@
+    for section in bears.keys():
+        new_bears[section] = tuple(
+            bear for bear in bears[section]
+            if {stdin.lower() for stdin in bear.use_stdin} & stdin_values)
```
@Makman2 
With `coala -B --filter-by use_stdin True`,
I'm getting the following error:

> type object 'VerilogLintBear' has no attribute 'use_stdin'

Bear object doesn't have use_stdin attribute. I'm not sure how to access a value from decorator's object.
Do you have any clues ?

Makman2 replied:
you shouldn't filter for `use_stdin`, this is an implementation detail of the bears. `use_stdin` is stored inside a dict of the bear decorated by `@linter`, but I'm not sure if it's scoped and thus not accessible. But like said, I wouldn't implement a filter for this^^ Maybe for the executable of a linter this might be useful, but not all bears are linter bears ;)

Makman2 replied:
Just look at `Linter.py`, although it's not so easy code ;)

shreyans800755 replied:
Makes sense. I don't think it would be much useful to end users since most of them won't even know what is `use_stdin` in the context.

Makman2 replied:
exactly :)

shreyans800755 commented:
@Makman2 
I've added the tests.
Currently the only function, I think, not working is when more than one `--filter-by` available. In that case, only later one is considered. I'll try to see if it can be fixed easily.

shreyans800755 commented:
@Makman2 
I've added `and` operator and couple tests in `FilterTest.py`.

Note: we'll have to wait for https://github.com/coala/coala/issues/4293 to get merged. Due to some issue, urllib3 is not working and causing Travis and appveyor build failure. (Surprisingly, circle CI is passing successfully)

Currently working register filter functionality.
What kind of interface are we looking for ?
Are we looking for end users functionality to make shortcuts for filter by registering it like the following example:
`coala -B --register-filter pylang language python`
`coala -B --filter-by pylang` or simply `coala -B pylang`

or am I misunderstanding the use case ?

userzimmermann commented:
@shreyans800755 The waiting for https://github.com/coala/coala/issues/4293 is over! https://github.com/coala/coala/pull/4294 has been merged :)

shreyans800755 commented:
@userzimmermann Thanks for the quick work. I'll just rebase.

Makman2 commented:
> Are we looking for end users functionality to make shortcuts for filter by registering it like the following example:
> coala -B --register-filter pylang language python
> coala -B --filter-by pylang or simply coala -B pylang
>
> or am I misunderstanding the use case ?

I meant registering filters from code side. The user shall not be able to register new filters (anyway way too complicated)

```
class LanguageFilter:
    ...

Filter.add_filter('language', LanguageFilter) # <-- example signature, we may want to move the specifier (here "language") to the filter class/callable/object
# or if LanguageFilter would contain the 'language' in itself, just
Filter.add_filter(LanguageFilter)

# we may built an instance of a filter with a certain feature-set, not sure if it makes sense in our case:
coala_filter = Filter()
coala_filter.add_filter(...)
coala_filter.add_filter(...)
# or maybe we support something like
coala_filter = Filter(LanguageFilter, AnotherFilter, AndAgainAnotherFilter, ...)
```

shreyans800755 commented:
@Makman2 
I'm still not able to find any good use case or intention behind this. Filter functionality is just to show the bears with specific functionality to users. I don't see how building a filter object will help us.
Can you please mention a use case where this might be useful ? And is it really in scope of this issue ? If not, may be we can file an issue and work can be done with respect to new issue.

Makman2 commented:
I don't have one, it was just a design intention. Feel free to ignore, I just wanted to give some input ;)

Makman2 commented:
Though we could actually try to automatically find filters in the same package, then you don't have to register it manually with your dict you currently use.

shreyans800755 commented:
Can you please elaborate a bit more ?

Makman2 commented:
Every filter we put into the same package where you define your main filter class handling everything can be autodetected maybe. Just an example:

```
# don't use star imports! I just do it because of simplicity!
from coalib.parsing.filters import *


def find_filters():
    for cls in dir():
        # Just an idea how we can identify filters using classes (I'm always using classes it seems...)
        if issubclass(cls, Filter):
            yield cls

class Filter:
    all_filters = set(find_filters) # <-- I know this needs to be a dict, but I hope you get the idea.
```

Actually when I'm thinking of it, this is quite difficult, nvm.

shreyans800755 commented:
I'll try to implement it.

Makman2 commented:
@shreyans800755 no really let's make it afterwards with a separate issue. This takes too much time in the first iteration.

shreyans800755 commented on coalib/coala.py:
```diff
@@ -59,26 +59,56 @@ def main(debug=False):
+                            filter, filter_args, all_bears)
+                    except Exception as ex:
+                        console_printer.print(ex)
+                        return 255
```
@Makman2 I'm returning 255 as error code in case the first argument of filter-by (type of filter) is not in the list. I'm not sure if there is a meaning of each return code in coala conventions. I tried to find if there's a list which describe the meaning of return codes, or list of constants/enums assigning a constant variable with return code, but didn't find it.
Can you please confirm if this return code is valid or any list of constant return code variable is available ?
Or if any other improvements are required, or is good to merge ?

Makman2 replied:
Yes each code has a meaning, and we have a list here: https://docs.coala.io/en/latest/Users/Exit_Codes.html

We still have some slots left, though maybe we should keep 255. Not sure^^

shreyans800755 replied:
> 2 - Invalid arguments were passed to coala in the command line.

error code 2 seems more suitable to me.
I'll update it.

shreyans800755 commented:
@Makman2 Are we looking for any other changes ? Or is this good to merge ?

Makman2 commented on coalib/coala.py:
```diff
@@ -59,26 +59,56 @@ def main(debug=False):
+                    "Use '--filter-by language ...' option instead")
+                if args.filter_by is None:
+                    args.filter_by = []
+                filter = ['language']
```
could you maybe alter the variable name a bit? `filter` is a built-in function, we shouldn't override it^^

Makman2 commented on coalib/coala.py:
```diff
@@ -59,26 +59,56 @@ def main(debug=False):
+                    filter_args = filter_by[1:]
+                    try:
+                        all_bears = Filter.get_filtered_bears(
+                            filter, filter_args, all_bears)
```
hm I dislike the name `all_bears`, as it's not all of them ;) Maybe `filtered_bears` fits better :+1:

Makman2 commented on coalib/coala.py:
```diff
@@ -59,26 +59,56 @@ def main(debug=False):
+                    try:
+                        all_bears = Filter.get_filtered_bears(
+                            filter, filter_args, all_bears)
+                    except Exception as ex:
```
This is too broad. Which exceptions can happen explicitly?

Makman2 commented on coalib/coala.py:
```diff
@@ -59,26 +59,56 @@ def main(debug=False):
+                           global_bears,
+                           args.show_description or args.show_details,
+                           args.show_details,
+                           console_printer)
```
the code in the else branch looks the same like in the normal "if" branch. The loop won't enter if there are no filters, so imo you can safely remove the `if`

Makman2 commented on coalib/coala_modes.py:
```diff
@@ -61,8 +59,10 @@ def mode_json(args, debug=False):
         try:
-            local_bears, global_bears = get_filtered_bears(
-                args.filter_by_language, LogPrinter())
+            from coalib.parsing.Filter import Filter
```
why not importing at top?

shreyans800755 replied:
Updated. Surprisingly its not able to resolve the reference of `Filter`. Logged at https://ci.appveyor.com/project/coala/coala/build/1.0.8117/job/78fcbqt9qp6atmyj#L507.

shreyans800755 replied:
Figured out, scope is not appropriate for `Filter`.

Makman2 commented on coalib/parsing/filters/CanDetectFilter.py:
```diff
@@ -0,0 +1,31 @@
+def filter_bears_by_can_detects(bears, detects):
+    local_bears, global_bears = bears
+    local_bears = filter_section_bears_by_can_detects(local_bears, detects)
+    global_bears = filter_section_bears_by_can_detects(
```
what about my suggestion to put this code into the filter class? Otherwise every filter needs to do this^^

shreyans800755 replied:
How about creating decorator for this ? And it will take care of this repetition.

shreyans800755 replied:
Figured out, scope is not proper for Filter.

shreyans800755 replied:
@Makman2  I've added implementation with class. Now, to add new filter, new class has to be derived from `FilterBase` and implement an abstract method.

shreyans800755 commented:
@Makman2 Can you please review this ?

Makman2 commented on coalib/parsing/filters/FilterBase.py:
```diff
@@ -0,0 +1,27 @@
+    def overrider(method):
+        assert(method.__name__ in dir(interface_class))
+        return method
+    return overrider
```
Hm that makes me remember java :D Usually not bad this kind of contract programming, but I think we should discuss that in a separate issue if we really want that. Until then I wouldn't use this as this is inconsistent with the rest of the codebase^^

shreyans800755 replied:
This is just to impose more restrictions to reduce errors and make the code a bit more self-documented.

We can remove this too, if not inconsistent. I've tried removing it in local and everything works fine. I can push changes if you want.

shreyans800755 replied:
Or may be this `overrides` decorator can be kept in some utility class/function so that whole codebase can use it easily.

shreyans800755 replied:
Its not about `ABC` and `@abstractmethod` that @Makman2 is discussing about.
Its mainly custom `@overrides` decorator.

userzimmermann commented on coalib/coala.py:
```diff
@@ -59,26 +60,49 @@ def main(debug=False):
+                # filters one by one provided as arguments
+                for filter_by in args.filter_by:
+                    filter_by_x = filter_by[0]
+                    filter_args = filter_by[1:]
```
You can also do just `filter_by_x, *filter_args = filter_by`

userzimmermann commented on coalib/coala.py:
```diff
@@ -59,26 +60,49 @@ def main(debug=False):
+                    args.filter_by = []
+                filter_by_x = ['language']
+                filter_by_x.extend(args.filter_by_language)
+                args.filter_by.append(filter_by_x)
```
You can simplify this to:

```python
                args.filter_by = args.filter_by or []
                args.filter_by += ['language'] + args.filter_by_language
```

shreyans800755 replied:
Have to keep `append` instead of `+=` as it is args and being used in the else part also. 
Rest of the suggestion is applied.

shreyans800755 commented:
@userzimmermann Pushed changes. Let me know if anything else to be changed.

Makman2 commented on coalib/parsing/filters/CanDetectFilter.py:
```diff
@@ -0,0 +1,21 @@
+
+        :param bears:       The dictionary of the sections as keys and list of
+                            bears as values.
+        :param args:        detects that bears are being filtered on.
```
detects -> Detects :)

Makman2 commented on coalib/parsing/filters/CanDetectFilter.py:
```diff
@@ -0,0 +1,21 @@
+
+    def filter_section_bears(self, bears, args):
+        """
+        Filters the bears by can_detect.
```
can_detect -->
```
``CAN_DETECT``
```

Makman2 commented on coalib/parsing/filters/CanFixFilter.py:
```diff
@@ -0,0 +1,21 @@
+
+    def filter_section_bears(self, bears, args):
+        """
+        Filters the bears by can_detect.
```
copy'n'paste gremlin sneaked in :P
-->
```
``CAN_FIX``
```

Makman2 commented on coalib/parsing/filters/CanFixFilter.py:
```diff
@@ -0,0 +1,21 @@
+
+        :param bears:       The dictionary of the sections as keys and list of
+                            bears as values.
+        :param args:        fixes that bears are being filtered on.
```
fixes -> Fixes

Makman2 commented on coalib/parsing/FilterHelper.py:
```diff
@@ -0,0 +1,31 @@
+                get_all_bears)
+            all_bears = get_all_bears(LogPrinter())
+        if FilterHelper.is_valid_filter(filter):
+            filter_object = FilterHelper.all_filters[filter]()
```
Use rather `@classmethod` instead of `@staticmethod` and do `cls.all_filters[filter]()`

(same for `is_valid_filter` above)

Makman2 commented on coalib/parsing/filters/CanDetectFilter.py:
```diff
@@ -0,0 +1,21 @@
+                tuple(bear for bear in bears[section]
+                      if {detect.lower()
+                          for detect in bear.CAN_DETECT} & detects)
+                for section in bears.keys()}
```
Do you need the section actually for filters?

Makman2 replied:
Btw: -->
```python
{section:
 tuple(bear for bear in bs
       if {detect.lower() for detect in bear.CAN_DETECT} & detects)
 for bs in bears.values()}
```
1. Iterating over `bears.keys()` is the same as iterating over just `bears`.
2. You just use the values, so just use `bears.values()` :)

shreyans800755 replied:
> Do you need the section actually for filters?

This is how `ConsoleInteraction.show_bears` takes parameters (with sections as keys of dict).

Makman2 replied:
ugh ... can we abstract the dict-building away into the `FilterBase` so we just we have to filter pure bears and don't need to build the dict? I think we are likely duplicating code for new filters with this^^
```python
def filter_bears(self, bears, args):
        local_bears, global_bears = bears
        filtered_local_bears = {section: self.filter_section_bears(bears) for section, bears in local_bears)
        filtered_global_bears = {section: self.filter_section_bears(bears) for section, bears in global_bears)
        return filtered_local_bears, filtered_global_bears
```

Makman2 replied:
And could you rename the `bears` parameter, to something like `sections_and_bears` or so? Just `bears` indicates that it's a pure collection of bears only :)

Makman2 commented on coalib/parsing/filters/CanDetectFilter.py:
```diff
@@ -0,0 +1,21 @@
+
+class CanDetectFilter(FilterBase):
+
+    def filter_section_bears(self, bears, args):
```
don't have a problem to keep it like this currently, but do we need `self`? Then you don't need to instantiate a filter object^^

Makman2 replied:
Nvm because you use base types^^

shreyans800755 replied:
Exactly. I had thought of making classmethod/staticmethod, but it didn't work with the obvious reason you specified.

Makman2 commented on coalib/parsing/filters/CanFixFilter.py:
```diff
@@ -0,0 +1,21 @@
+                tuple(bear for bear in bears[section]
+                      if {fix.lower()
+                          for fix in bear.CAN_FIX} & fixes)
+                for section in bears.keys()}
```
--> `bears.values()`, see above

Makman2 commented on coalib/parsing/filters/LanguageFilter.py:
```diff
@@ -0,0 +1,20 @@
+        return {section:
+                tuple(bear for bear in bears[section]
+                      if {lang.lower() for lang in bear.LANGUAGES} & languages)
+                for section in bears.keys()}
```
--> `bears.values()`, see above

shreyans800755 replied:
It will be `bears` and not `bears.values()` as I'm going through keys and not values for all three filters.

Makman2 commented on coalib/parsing/filters/__init__.py:
```diff
@@ -0,0 +1,5 @@
+"""
+This package holds filter objects. Filter objects are used to
+get the list of bears with specific properties.
+
```
redundant newline

Makman2 commented on coalib/parsing/filters/__init__.py:
```diff
@@ -0,0 +1,5 @@
+"""
+This package holds filter objects. Filter objects are used to
+get the list of bears with specific properties.
```
`the` -> `a`, sounds a bit nicer ;)

Makman2 commented on coalib/settings/ConfigurationGathering.py:
```diff
@@ -4,7 +4,7 @@
 
 from coalib.collecting.Collectors import (
-    collect_all_bears_from_sections, filter_section_bears_by_languages)
+    collect_all_bears_from_sections)
```
this does fit into a single line now :+1:

Makman2 commented on coalib/parsing/FilterHelper.py:
```diff
@@ -0,0 +1,31 @@
+
+    @staticmethod
+    def get_help_message(filter):
+        return '{} is invalid filter.\nAvailable filters: {}'.format(
```
--> is **an** invalid filter.

Makman2 replied:
I would rather avoid `\n`s in tracebacks, they usually distort the output layout. But up to you^^

Makman2 replied:
Maybe you want to use `{!r}` so the filter string gets enclosed inside quotes?

Makman2 commented on coalib/parsing/FilterHelper.py:
```diff
@@ -0,0 +1,31 @@
+    @staticmethod
+    def get_help_message(filter):
+        return '{} is invalid filter.\nAvailable filters: {}'.format(
+            filter, ', '.join(sorted(FilterHelper.all_filters.keys())))
```
you don't need `.keys()` :)

Makman2 commented:
I'm still a bit unsure about using filter objects as we don't store any additional information except the filter function itself currently. The base class could be inlined into `FilterHelper`, so we don't need to override anything any more and we can use pure functions.

But the design is okay for now imo :)

shreyans800755 commented:
@Makman2 

> The base class could be inlined into FilterHelper, so we don't need to override anything any more and we can use pure functions.

I think, this way, we've got more restrictions and proper structure. If someone wants to add a filter, one just have to inherit from FilterBase and implement abstract method.
This way, I've tried to follow strategy design pattern, since we've got different implementation of filters and can choose one at a time to get the output.
IMO, by inlining FilterBase into FilterHelper, it would be highly coupled, and if any changes come into filters structure, it might become more difficult to incorporate new structure. IMO, losely coupled approach is better and more maintainable.

Makman2 commented:
> and if any changes come into filters structure

You have something in mind that might change in future? :D
What I only have in mind is that with the next-gen core there will be no distinction between Local and GlobalBears any more^^

Anyway, sounds fine :D

Makman2 commented on coalib/parsing/DefaultArgParser.py:
```diff
@@ -187,6 +187,10 @@ def default_arg_parser(formatter_class=None):
 
     outputs_group.add_argument(
+        '--filter-by', action='append', nargs='+',
+        help='filter --show-bears by the filter given as argument.')
```
Maybe we should list available filters here or somewhere else and refer?

shreyans800755 replied:
Write now,
`coala -B --filter-by` gives default error of `Usage : all the possible options...`.
But `coala -B --filter-by unknown` gives the following error:
`'unknown' is an invalid filter. Available filters: can_detect, can_fix, language`

IMO, this is sufficient. We don't need to show which filters are available in former query.

Makman2 replied:
Though sometimes people don't use the feature when they don't know whether typing something wrong gives a better message (and I would count to those people).

Makman2 replied:
If I get immediate help, then I immediately use that :D

shreyans800755 replied:
@Makman2 That makes sense. Currently there's no direct way of knowing which filters are available unless user tries to filter bears with some wrong filter name.
One solution I can think of is to add one more option called `--show-filters` in `--show-bears` which will display all the available filters.
I'm confused b/w two of the following options:
`coala -B --show-filters`
or
`coala -B show-filters`

What's your opinion ?

Makman2 replied:
why not embedding that into the argparser help? You can just fetch available filters from `FilterHelper` and construct a message :)

Makman2 commented on coalib/parsing/filters/FilterBase.py:
```diff
@@ -0,0 +1,20 @@
+    def get_filtered_bears(self, bears, args):
+        if args:
+            bears = self.filter_bears(bears, args)
+        return bears
```
up to you: I think it's not really necessary to have a function `get_filtered_bears` which just checks for the presence of arguments and then invokes another function in that case. I think it's simple enough to inline it into `filter_bears`. External calls look also a bit nicer with `filtered = filter_bears(...)` instead of `filtered = get_filtered_bear(...)` (which indicates a bit for me that the bears were already filtered and stored somewhere :D)

Makman2 commented on coalib/coala.py:
```diff
@@ -59,26 +60,45 @@ def main(debug=False):
                        console_printer)
-
             return 0
+
```
unnecessary change :)

Makman2 commented on coalib/coala.py:
```diff
@@ -59,26 +60,45 @@ def main(debug=False):
                        args.show_description or args.show_details,
                        args.show_details,
                        console_printer)
-
```
unnecessary change :)

Makman2 commented on coalib/coala.py:
```diff
@@ -59,26 +60,45 @@ def main(debug=False):
+                # Each iteration of the following loop applies
+                # filters one by one provided as arguments
+                for filter_by in args.filter_by:
+                    filter_by_x, *filter_args = filter_by
```
wahhhh this is possible? I haven't known that there's such a syntax! :O

shreyans800755 replied:
Kudos to @userzimmermann . He suggested this at https://github.com/coala/coala/pull/4240#discussion_r122594079.

userzimmermann replied:
@Makman2 That's Python 3 ;)

Makman2 commented on coalib/coala.py:
```diff
@@ -59,26 +60,45 @@ def main(debug=False):
+            if args.filter_by:
+                # Each iteration of the following loop applies
+                # filters one by one provided as arguments
+                for filter_by in args.filter_by:
```
what if `filter-by` is empty? I think it would crash below because of an unpacking error^^

shreyans800755 replied:
Arg parser takes care of this.
`coala -B --filter-by` gives error thrown by arg parser with `Usage` message.

Makman2 commented on coalib/settings/ConfigurationGathering.py:
```diff
@@ -3,8 +3,7 @@
 
-from coalib.collecting.Collectors import (
-    collect_all_bears_from_sections, filter_section_bears_by_languages)
+from coalib.collecting.Collectors import (collect_all_bears_from_sections)
```
no braces necessary around import :D

shreyans800755 commented:
> You have something in mind that might change in future?

I don't have anything in mind at the moment. But with the changes in core design through aspects and all, I though there be some change in future for filters too. Just thinking out loud, and not much consolidated idea.

shreyans800755 commented:
@Makman2 
Added filters in help message.

One analysis (a bit out of context for this PR):
I pushed my changes here, and got few errors from gitmate-bot regarding PEP8 styling (Sadly I forgot to do it in local before pushing). But my travis-ci job also failed due to this reason. Build and tests succeeded, but the status of build was **red**. That's because of the command `coala --non-interactive` run in travis. I'm curious if its really required to do the same thing multiple times at different places. IMO, job of travis should be only limited to test the code. (Appveyor passed successfully though, as it doesn't have this command)

Makman2 commented:
>  I'm curious if its really required to do the same thing multiple times at different places.

Yeah you are right, if that's the case, we should encourage to optimize our build times. Feel free to file an issue :+1:

> Apart from that, Circle CI is giving time out. I've faced this situation multiple times. Although it passes on rebuild. Is there any issue filed for this ?

Do you have a link to a timeouted build?

Makman2 commented on coalib/coala.py:
```diff
@@ -59,11 +60,35 @@ def main(debug=False):
+                # Since it is already covered by filter_by option.
+                log_printer.warn(
+                    "'--filter-by-language ...' option is deprecated. "
+                    "Use '--filter-by language ...' option instead")
```
missing period at end of sentence :D

Makman2 commented on coalib/coala.py:
```diff
@@ -59,11 +60,35 @@ def main(debug=False):
+                log_printer.warn(
+                    "'--filter-by-language ...' option is deprecated. "
+                    "Use '--filter-by language ...' option instead")
+                args.filter_by = args.filter_by or []
```
isn't this one always a list because of the argument parser?
Maybe you can omit this statement?

shreyans800755 replied:
If args.filter_by is None, then it will be empty list after this statement.

This is to cover the following case:
`coala -B --filter-by-language java`

In this case, we first build `args.filter_by` from `args.filter_by_language`. And then process it, to make it simple. And it will also make it easier to remove `if args.filter_by_language:` in future as it will be deprecated after adding this generic filter processor.

Makman2 replied:
Can't you set a default for `--filter-by` to use an empty list?

shreyans800755 replied:
I can set default value while adding it in `DefaultArgParser.py`. But this is causing test failures in the following tests:
`tests/output/ConfWriterTest.py::ConfWriterTest::test_append`
`tests/output/ConfWriterTest.py::ConfWriterTest::test_write`

Because it assumes that config doesn't have `filter_by` option as it is not specified in this test.

Similarly, following two are also failing:
tests/settings/ConfigurationGatheringTest.py::ConfigurationGatheringTest::test_back_saving
tests/settings/ConfigurationGatheringTest.py::ConfigurationGatheringTest::test_gather_configuration

shreyans800755 replied:
We can fix it by adding `filter_by` attribute though.
IMO, its better to keep it this way, since this statement is anyways part of condition which is to be deprecated.

Makman2 replied:
Yeah in this case let's keep using `None` :+1:

Makman2 replied:
Ah maybe to make it more clear -->
```
args.filter_by = [] if args.filter_by is None else args.filter_by
```
Up to you, I'm also fine with the short-circuitance-using expression :)

shreyans800755 replied:
Good enough. Makes it more readable.

Makman2 commented on coalib/coala.py:
```diff
@@ -59,11 +60,35 @@ def main(debug=False):
+                        continue
+                    try:
+                        filtered_bears = FilterHelper.get_filtered_bears(
+                            filter_by_x, filter_args, filtered_bears)
```
This is just a design consideration, you don't have to implement it (it's actually not so performant, just for the concept).

Mathematically speaking, because we apply a filter on a set of bears, and especially in a sequence, the invocation order of filters may define different results. A filter usually applies to a single bear, so we don't need to pass a filter the complete set of bears to filter. To ensure that the order of invocation and actual set of bears is independent from the result, we could invoke a single function that just says "yeah this bear shall be taken into the results -> True" or "no the bear should be filtered out -> False".
This kind of principle btw exists in many languages (also in Python with list comprehensinos), they usually take in a lambda that defines when an object shall be kept or not with a boolean value.

shreyans800755 replied:
Here, the invocation order of filters won't give different results.
If I understood your comment correctly,
In your opinion:
`coala -B --filter-by language c --filter-by can_fix syntax`
and
`coala -B --filter-by can_fix syntax --filter-by language c`
can give different results.

But, as per the implementation, the result is nothing but a set of bears containing intersection of all the bears filtered out from each filters individually. However, implementation is not directly taking intersection, but it doing something like folding. But the essence is same. We just want to show the intersection of all the filters.

Correct me if I'm wrong.

Makman2 commented on coalib/coala.py:
```diff
@@ -59,11 +60,35 @@ def main(debug=False):
+                            filter_by_x, filter_args, filtered_bears)
+                    except NotImplementedError:
+                        logging.exception(filter_by_x +
+                                          ' filter is not implemented')
```
Missing period :)

Makman2 commented on coalib/parsing/DefaultArgParser.py:
```diff
@@ -187,6 +188,12 @@ def default_arg_parser(formatter_class=None):
     outputs_group.add_argument(
+        '--filter-by', action='append', nargs='+',
+        help='filters `--show-bears` by the filter given as argument.'
+             'Available filters: {}'.format(
```
missing space in string after `argument.` / before `Available` :)
Renders as
```
... by the filter given as argument.Available filter: ...
```
currently

Makman2 commented on coalib/parsing/FilterHelper.py:
```diff
@@ -0,0 +1,38 @@
+
+    @classmethod
+    def get_all_filters_str(cls, sep=', '):
+        return sep.join(sorted(cls.all_filters))
```
I think you should inline this into the `argparse` code above, imo this is usually application specific code, having a function here is not so modular imo :+1:

shreyans800755 replied:
Its also being used in `FilterHelper.get_help_message` function. So, I think it makes sense not to repeat it at bother places.

Makman2 replied:
Hmm. Okay that makes sense.

Although I imagine when I would now use your Filter stuff, I actually wouldn't look into your class when I need to build some help strings ;) Especially I wouldn't understand what they are exactly producing as output and where they are especially useful for, and then I'm usually faster writing some quick expression myself (thank you Python :D), using `all_filters`.

Makman2 commented on coalib/coala.py:
```diff
@@ -59,11 +60,35 @@ def main(debug=False):
+                for filter_by in args.filter_by:
+                    filter_by_x, *filter_args = filter_by
+                    if len(filter_args) == 0:
+                        continue
```
you already handle this case in `FilterHelper`, I think you can safely omit it here :)

shreyans800755 commented:
@Makman2 

> Do you have a link to a timeouted build?

https://circleci.com/gh/coala/coala/13182

Makman2 commented:
Ahhh... `language-check`... We have an issue for that (can't find it right now). It was always a pain in our builds ;D

Makman2 commented on coalib/parsing/FilterHelper.py:
```diff
@@ -0,0 +1,38 @@
+
+
+class FilterHelper:
+    all_filters = {'language': LanguageFilter,
```
hm just thinking, maybe `available_filters` is a better name? What do you think?

userzimmermann commented on coalib/coala.py:
```diff
@@ -59,11 +60,34 @@ def main(debug=False):
+                    "'--filter-by-language ...' option is deprecated. "
+                    "Use '--filter-by language ...' option instead.")
+                args.filter_by = [] if args.filter_by is None else \
+                    args.filter_by
```
Continuation of https://github.com/coala/coala/pull/4240#discussion_r122601529

In this solution the `else` part is obsolete. Then it would be better to keep:

```python
if args.filter_by is None:
    args.filter_by = []
```

But IMO it's still sufficient enough to just do:

```python
args.filter_by = args.filter_by or []
```

If it's already an empty `list` it doesnt' really hurt ;)

shreyans800755 replied:
`args.filter_by = args.filter_by or []` is definitely sufficient. Main concern is readability. I think, as you mentioned,
`if args.filter_by is None:
    args.filter_by = []`
is better option in terms of readability.

Makman2 replied:
hm the two-liner if is also fine with me, actually it's slightly more performant^^ But up to you @shreyans800755

Makman2 commented on coalib/coala.py:
```diff
@@ -59,11 +60,34 @@ def main(debug=False):
+                    "Use '--filter-by language ...' option instead.")
+                args.filter_by = [] if args.filter_by is None else \
+                    args.filter_by
+                args.filter_by.append(['language'] + args.filter_by_language)
```
shouldn't it be `extend`/`+=`? `append` would place the list itself as an item into `args.filter_by`, so in the end you have
`[['language', 'python']]`

Makman2 replied:
I'm not sure at least that `argparse` does it like this^^

shreyans800755 replied:
If we use `extend`, it would be:
`['language', 'python']` and not `[['language', 'python']]`. `extend` essentially appends all the elements present in the list given as argument. But, we want to append it as list and not as individual elements.

Regarding, `+=`, it actually creates new object. Since, `args.filter_by` is not our own variable, and its created by `argparse`, its not advised to change its reference. It might to lead to unexpected behavior. And there's no advantage of it.

Makman2 replied:
> Regarding, `+=`, it actually creates new object.

Nope it doesn't, it's in-place :)

```
>>> a=[1,2,3]
>>> id(a)
140015517382472
>>> a+=[4,5]
>>> id(a)
140015517382472
```

But okay so having a twice-nested list is intended. Does argparse create the same, e.g. when doing
```
--filter-by language python C --filter-by can_fix all
```
Does it create `[['language', 'python', 'C'], ['can_fix', 'all']]`?

shreyans800755 replied:
Yeah, argparse create the list as you mentioned.

Makman2 commented on coalib/coala.py:
```diff
@@ -59,11 +60,34 @@ def main(debug=False):
+                                          ' filter is not implemented.')
+                        return 255
+                    except AttributeError as ex:
+                        console_printer.print(ex)
```
use `logging.exception` instead

Makman2 replied:
When did this error occur again? I already forgot... maybe you should place a comment :/

Makman2 replied:
ahhh you raise an `AttributeError` for non-existing filters. I would consider implementing a new exception, like `InvalidFilterException`. An attribute error is so general that it might occur everywhere, and possible bugs would slip through.

Makman2 commented on coalib/coala_modes.py:
```diff
@@ -61,8 +59,10 @@ def mode_json(args, debug=False):
+            from coalib.parsing.FilterHelper import FilterHelper
+
+            local_bears, global_bears = FilterHelper.get_filtered_bears(
+                'language', args.filter_by_language)
```
also here you are using the old setting again^^

shreyans800755 replied:
I think we need a flag to get the list of languages apart from `filter_by_language` (`-l`). May be `--language`. Or we gotta change the meaning of `-l` to language, as filter_by_language doesn't always make sense.

This condition is not making sense to me (`if args.show_bears`). We are calling the function `mode_json` from coala if `args.json` is true. IMO, we should do processing using normally, and at the time of dumping data to stdout (using console_printer), we should check if the output format is chosen by user to some specific format(`json` in this case).
This seems like repeatation of code at the moment.

shreyans800755 replied:
At the moment, there's no other way of providing languages apart from `filter_by_language`. So, to use `filter_by` over here, we need to repeat the same logic as present in coala.py

Makman2 replied:
hmm kk. Let's file an issue for that after merge to find some neat idea without duplicating code :+1:

Makman2 commented on coalib/parsing/FilterHelper.py:
```diff
@@ -0,0 +1,38 @@
+    @classmethod
+    def get_help_message(cls, filter):
+        return '{!r} is an invalid filter. Available filters: {}'.format(
+            filter, cls.get_all_filters_str())
```
as you only use this message inside the error, I would rather inline it. Nobody would call `get_help_message` again ;)

Makman2 commented on tests/parsing/FilterTest.py:
```diff
@@ -0,0 +1,128 @@
+                coala.main, 'coala', '-B', '--filter-by', 'unknown', 'arg1')
+            self.assertEqual(retval, 2)
+            self.assertEqual(
+                stdout, FilterHelper.get_help_message('unknown') + '\n')
```
I would rather write the messages out, testing by expected output, not by expected algorithm behaviour ;)

(applies everywhere in this test ;) )

Makman2 commented on coalib/parsing/FilterHelper.py:
```diff
@@ -0,0 +1,38 @@
+            return all_bears
+        if cls.is_valid_filter(filter):
+            filter_object = cls.available_filters[filter]()
+            return filter_object.filter_bears(all_bears, args)
```
have you read my comment before? To execute filters for each section, so the users don't need to handle them?
For filter writers it would be really confusing, e.g. I would really ask myself "why do I need this?", and especially "what shall I do with it?"

shreyans800755 replied:
I've already removed a unnecessary function from `FilterBase`. Anyways, this is not something filter writers have to worry about. They just gotta implement `filter_section_bears` function for the new filter. This is part of helper class and not base class of filters.

Makman2 commented:
Some relevant discussion in gitter, just for reference:
> @shreyans800755 hm, regarding filtering, I've been thinking again about using a function that returns `True`/`False` instead of the completely filtered set
and I'm quite tempted to it, because for example the CAN_FIX filter, would decrease from 5 lines to 1!
> ```
> return any(fix.lower() in bear.CAN_FIX for fix in args)
> ```
>
> instead of
> ```
>        fixes = {fix.lower() for fix in args}
>        return {section:
>                tuple(bear for bear in bears[section]
>                      if {fix.lower()
>                          for fix in bear.CAN_FIX} & fixes)
>                for section in bears}
> ```
> (oh it's even 6 lines --> 1) :3
> or to keep your intersection behaviour:
> ```
> return bool({fix.lower() for fix in args} & bear.CAN_FIX)
> ```
> (trivia: both 1-liner versions perform differently regarding performance, depending on the number of arguments passed and whether they are in bear.CAN_FIX or not, just tested it :wink: )
> @shreyans800755 using those functions would make those filter also way easier to test. Then you just need to make a few little tests invoking coala, so you can see whether filter functionality is correctly executed, and then you can battle-test the filters separately
> \+ I'm still not very convinced using a FilterBase class :confused: filters don't hold a state (at least not now), so classes would be not the right design choice imo. I would agree on keeping them now, but I would raise an issue discussing that again :wink: (fyi :smiley:)
> especially there's a division-of-work problem, what does the FilterHelper, FilterBase and our coala implementation do? What are there duties?
when I would write now a filter, I would really wonder why I need to use base class in python for that. Usually I would expect to write a simple lambda or so^^
instead of classes you could maybe use decorators, just as an idea for later :thumbsup:

shreyans800755 commented on coalib/coala.py:
```diff
@@ -74,11 +98,8 @@ def main(debug=False):
-            local_bears, global_bears = get_filtered_bears(
-                args.filter_by_language, log_printer)
+            local_bears, global_bears = FilterHelper.get_filtered_bears(
+                'language', args.filter_by_language)
```
Analyzing it properly, it should ideally be `args.show_capabilities` since languages are given as the arguments of option `show_capabilities` as per tests.
So, `coala -p java` works fine. No need of `coala -p -l java`.

shreyans800755 commented:
@Makman2 
Regarding usage of decorator, As far as I can remember, I had suggested this alternative, but moved further with this approach. But definitely worth giving a though. Like you said, may be we should discuss it over some other issue.
Apart from that, there are multiple parameters involved in here, like local/global bears, sections, etc. Some of which are to be deprecated in future, and will be faded away. Due to these parameters, design is quite complex at the moment as these require more and more processing at each and every step.

shreyans800755 commented:
@Makman2 I had not really thought of using boolean values for filters. That's actually a good idea. I wasn't aware of the function `any()` :smile:

shreyans800755 commented:
From gitter at https://gitter.im/coala/coala?at=5955421c11755ab05636c3d0:

> @Makman2 May be a separate issue can be filed, which can make these objects light weight, and only contain necessary info like name, description, etc. It will also improve the speed. A function similar to get_all_bears, but has only necessary info, and we can make it independent of local/global bears as that's not a concern in most functionalities like filter or just showing the bear's info.
Apart from that, may be one more filter can be added to filter bears by name. One use case is that, after getting name from web or somewhere else, user should be able to get description on cli. Something like coala -B --show-description --filter-by name cppcheck would give the description of cppcheckbear. Just an idea. Not sure whether it should be part of this PR or not.

shreyans800755 commented:
@Makman2 
Removed inheritance approach, and added decorator approach. Also shifted main logic in the decorator, so that it need not be repeated whenever we want to add a new filter.
Just one concern I can see here:
Right now, filter function looks like this: `xxx_filter(bears, args)`. I couldn't enforce the return type to be `bool` for any filter. This will not cause any big issues as it will be the job of the developer writing new filters, and it will never reach to end user. But its just something good to have IMO. (Or may be I'm just using too much of static typed languages(C++) :smile: )
Let me know if you've any suggestions for this.

Makman2 commented:
It's Python, we are all adults here, don't enforce types ;) (if it's really not necessary ;D)

Actually I also like type safety, in many cases it's - yeah- just safe :D

Makman2 commented on coalib/parsing/FilterHelper.py:
```diff
@@ -0,0 +1,37 @@
+
+
+class InvalidFilterException(Exception):
+    pass
```
We should put the exception into an own file ;)

Makman2 commented on coalib/parsing/filters/CanDetectFilter.py:
```diff
@@ -0,0 +1,14 @@
+    :return:            True if this bear is to be displayed for given
+                        args, False otherwise
+    """
+    return {detect.lower() for detect in bear.CAN_DETECT} & args
```
not sure, maybe we should explicitly "boolify" the value?
```
return bool({detect.lower() for detect in bear.CAN_DETECT} & args)
```
I don't know. Up to you. There's actually no real reason to do this, except that it's maybe slightly better readable.

shreyans800755 replied:
Its not really required, as this function is used as if condition. So, it is by default converted to bool. Still I added this, just to make it more readable as you mentioned.

Makman2 commented on coalib/parsing/filters/CanFixFilter.py:
```diff
@@ -0,0 +1,14 @@
+    :param bear:        Bear object
+    :param args:        Set of fixes on which bear is to be filtered
+    :return:            True if this bear is to be displayed for given
+                        args, False otherwise
```
missing periods at end of sentences :)
\+
Enclose True and False into double backticks, as these are code terms ;)
(applies for all filter docstrings)

Makman2 commented on coalib/parsing/filters/Filter.py:
```diff
@@ -0,0 +1,15 @@
+        global_bears = filter_section_bears(global_bears, args)
+        return local_bears, global_bears
+
+    def filter_section_bears(bears, args):
```
maybe put `filter_section_bears` before `filter_wrapper`? As you use it above "before it's defined" regarding reading order. Up to you :+1:

Makman2 commented on coalib/parsing/filters/Filter.py:
```diff
@@ -0,0 +1,15 @@
+def filter(filter_applier):
+    def filter_wrapper(bears, args):
```
use `functools.wraps` to make the old function signature appear for debugging purposes

shreyans800755 commented:
@Makman2 Pushed changes as requested.

Makman2 commented on coalib/parsing/FilterHelper.py:
```diff
@@ -0,0 +1,35 @@
+from coalib.parsing.filters.CanFixFilter import can_fix_filter
+
+
+class FilterHelper:
```
hm kinda interesting: You never use an instance of this class :D (makes also no sense as there are no class members :D)

Up to you: Maybe moving all functions to module scope^^

shreyans800755 replied:
I think this way it looks neater.

Makman2 commented on coalib/parsing/filters/CanDetectFilter.py:
```diff
@@ -0,0 +1,14 @@
+    :param bear:        Bear object.
+    :param args:        Set of detects on which bear is to be filtered.
+    :return:            ``True`` if this bear is to be displayed for given
+                        args, ``False`` otherwise.
```
please align docs without leaving so much space :+1:

Makman2 replied:
> ``True`` if this bear is to be displayed for given args, ``False`` otherwise.

Maybe rephrase this into something more general^^

> ``True`` if this bear matches the criteria inside ``args``, ``False`` otherwise.

Up to you :+1:

Makman2 replied:
Applies to all filters :+1:

Makman2 commented on coalib/parsing/filters/CanFixFilter.py:
```diff
@@ -0,0 +1,14 @@
+    :param bear:        Bear object.
+    :param args:        Set of fixes on which bear is to be filtered.
+    :return:            ``True`` if this bear is to be displayed for given
+                        args, ``False`` otherwise.
```
Same here with alignment

Makman2 commented on coalib/parsing/filters/CanFixFilter.py:
```diff
@@ -0,0 +1,14 @@
+@filter
+def can_fix_filter(bear, args):
+    """
+    Filters the bears by ``LANGUAGES``.
```
copy'n'paste worm :D

shreyans800755 replied:
Ugh...

Makman2 commented on coalib/parsing/filters/CanDetectFilter.py:
```diff
@@ -0,0 +1,14 @@
+@filter
+def can_detect_filter(bear, args):
+    """
+    Filters the bears by ``LANGUAGES``.
```
here also, copy'n'paste worm :D

Makman2 commented on coalib/coala.py:
```diff
@@ -59,11 +60,31 @@ def main(debug=False):
+                # Since it is already covered by filter_by option.
+                log_printer.warn(
+                    "'--filter-by-language ...' option is deprecated. "
+                    "Use '--filter-by language ...' option instead.")
```
Use '--filter-by language ...' ~~option~~ instead.

Makman2 commented on coalib/coala.py:
```diff
@@ -59,11 +60,31 @@ def main(debug=False):
+                # filter_by_language option should be deprecated soon.
+                # Since it is already covered by filter_by option.
+                log_printer.warn(
+                    "'--filter-by-language ...' option is deprecated. "
```
'--filter-by-language ...' ~~option~~ is deprecated.

Makman2 commented on coalib/coala.py:
```diff
@@ -59,11 +60,31 @@ def main(debug=False):
+                            filter_by_x, filter_args, filtered_bears)
+                    except InvalidFilterException as ex:
+                        # If filter is not available
+                        logging.exception(ex)
```
Does this show a traceback to the user? As it's common that a user might mistype a filter name, we only should display the exception message.

shreyans800755 replied:
Yeah, it actually prints traceback to the user. I'll just use `console_printer.print` to print the message.

Makman2 commented on coalib/parsing/filters/CanDetectFilter.py:
```diff
@@ -0,0 +1,14 @@
+    Filters the bears by ``LANGUAGES``.
+
+    :param bear:        Bear object.
+    :param args:        Set of detects on which bear is to be filtered.
```
`bear` inside double backticks, as you refer to the parameter.

Makman2 replied:
Maybe `Set of detectable issue types`? Up to you :+1:

Makman2 commented on coalib/parsing/filters/CanFixFilter.py:
```diff
@@ -0,0 +1,14 @@
+    Filters the bears by ``LANGUAGES``.
+
+    :param bear:        Bear object.
+    :param args:        Set of fixes on which bear is to be filtered.
```
`bear` inside double backticks, as you refer to the parameter.

Makman2 replied:
Maybe `Set of fixable issue types`? Up to you again :+1:

Makman2 commented on coalib/parsing/filters/Filter.py:
```diff
@@ -0,0 +1,19 @@
+from functools import wraps
+
+
+def filter(filter_applier):
```
`filter_applier` -> `filter_function`

Makman2 commented on coalib/parsing/filters/Filter.py:
```diff
@@ -0,0 +1,19 @@
+
+
+def filter(filter_applier):
+    def filter_section_bears(bears, args):
```
you should rename the `bears` parameter into something that does fit more. It contains also section names as keys^^

shreyans800755 replied:
How about `section_to_bears_dict` or may be just `section_to_bears` ?

Makman2 commented on coalib/parsing/filters/__init__.py:
```diff
@@ -0,0 +1,4 @@
+"""
+This package holds filter objects. Filter objects are used to
+get a list of bears with specific properties.
```
maybe filter objects --> filter functions?

Makman2 commented on coalib/parsing/filters/CanDetectFilter.py:
```diff
@@ -0,0 +1,15 @@
+    :param args:    Set of detectable issue types on which ``bear`` is to be
+                    filtered.
+    :return:        ``True`` if this bear matches the criteria inside args,
+                    ``False`` otherwise.
```
still indentation:
```
:param bear: Bear object.
:param args: Set of...
:return:     ``True`` if this bear...

Makman2 replied:
Same applies below

sils commented on coalib/coala.py:
```diff
@@ -59,11 +60,31 @@ def main(debug=False):
+            filtered_bears = get_all_bears(log_printer)
+            if args.filter_by_language:
+                # filter_by_language option should be deprecated soon.
+                # Since it is already covered by filter_by option.
```
this comment doesn't really add info to what's in the warning below so IMO it can just be removed

Makman2 commented on tests/coalaTest.py:
```diff
@@ -185,7 +186,8 @@ def test_show_capabilities_with_supported_language_debug(self):
             import_fn.side_effect = VersionConflict('msg1', 'msg2')
-            retval, stdout, stderr = execute_coala(coala.main, 'coala', '-B')
+            retval, stdout, stderr = execute_coala(
+                coala.main, 'coala', '-B', '-l ')
```
why the space after `-l`?

Makman2 replied:
btw what does `-l` mean?

shreyans800755 replied:
In fact, `-l`  is not required. Can't recall why I added it. Removing.

Makman2 commented:
ack 498c731

Makman2 commented:
Let's do it! :D
@rultor merge

Makman2 commented:
@rultor merge

rultor commented:
> @rultor merge

@makman2 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/12436-316850035)

rultor commented:
> @rultor merge

@makman2 Done! FYI, the full log is [here](http://www.rultor.com/t/12436-316850035) (took me 2min)
